File: src/main/java/cspfj/generator/constraint/AddGenerator.java
package cspfj.generator.constraint;

import static com.google.common.collect.Sets.newTreeSet;

import java.util.NoSuchElementException;
import java.util.SortedSet;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterables;
import com.google.common.primitives.Ints;

import cspfj.constraint.Constraint;
import cspfj.constraint.semantic.Add;
import cspfj.constraint.semantic.Eq;
import cspfj.exception.FailedGenerationException;
import cspfj.problem.BitVectorDomain;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspom.constraint.CSPOMConstraint;

public final class AddGenerator extends AbstractGenerator {

    public AddGenerator(final Problem problem) {
        super(problem);
    }

    @Override
    public boolean generate(final CSPOMConstraint constraint)
            throws FailedGenerationException {
        final Variable result;
        final Variable v0;
        final Variable v1;
        if ("sub".equals(constraint.description())) {
            result = getSolverVariable(constraint.getVariable(1));
            v0 = getSolverVariable(constraint.getVariable(0));
            v1 = getSolverVariable(constraint.getVariable(2));
        } else if ("add".equals(constraint.description())) {
            result = getSolverVariable(constraint.getVariable(0));
            v0 = getSolverVariable(constraint.getVariable(1));
            v1 = getSolverVariable(constraint.getVariable(2));
        } else {
            throw new IllegalArgumentException("Cannot handle " + constraint);
        }

        try {
            final Variable nullVariable = Iterables.getOnlyElement(Iterables
                    .filter(ImmutableList.of(result, v0, v1), NULL_DOMAIN));

            if (nullVariable == result) {

                final SortedSet<Integer> values = newTreeSet();
                for (int i : v0.getDomain().allValues()) {
                    for (int j : v1.getDomain().allValues()) {
                        values.add(i + j);
                    }
                }
                result.setDomain(new BitVectorDomain(Ints.toArray(values)));

            } else if (nullVariable == v0) {

                v0.setDomain(new BitVectorDomain(generateValues(result, v1)));

            } else if (nullVariable == v1) {

                v1.setDomain(new BitVectorDomain(generateValues(result, v0)));

            } else {

                throw new IllegalStateException();

            }
        } catch (NoSuchElementException e) {
            // All variables are defined, continue...
        } catch (IllegalArgumentException e) {
            // More than 1 variable is undefined
            return false;
        }

        final Constraint generated;
        if (result.getDomainSize() == 1) {
            generated = new Eq(-1, v0, result.getFirstValue(), v1);
        } else if (v0.getDomainSize() == 1) {
            generated = new Eq(1, v1, v0.getFirstValue(), result);
        } else if (v1.getDomainSize() == 1) {
            generated = new Eq(1, v0, v1.getFirstValue(), result);
        } else {
            generated = new Add(result, v0, v1);
        }
        addConstraint(generated);
        return true;

    }

    private static int[] generateValues(final Variable result,
            final Variable var) {
        final SortedSet<Integer> values = newTreeSet();
        for (int i : result.getDomain().allValues()) {
            for (int j : var.getDomain().allValues()) {
                values.add(i - j);
            }
        }

        return Ints.toArray(values);
    }

}

File: src/main/scala/cspfj/generator/constraint/AddGenerator.scala
package cspfj.generator.constraint;

import cspfj.constraint.semantic.{ Eq, Add }
import cspfj.constraint.Constraint
import cspfj.{ Variable, Problem, IntDomain }
import cspom.constraint.CSPOMConstraint
import cspom.constraint.FunctionalConstraint

final class AddGenerator(problem: Problem) extends AbstractGenerator(problem) {

  override def generateFunctional(constraint: FunctionalConstraint) = {

    val Seq(result, v0, v1) = (constraint.description match {
      case "sub" => Seq(constraint.arguments(0), constraint.result, constraint.arguments(1))
      case "add" => constraint.result +: constraint.arguments
      case _ => throw new IllegalArgumentException("Cannot handle " + constraint)
    }) map cspom2cspfj

    if (Seq(result, v0, v1) filter { _.dom == null } match {
      case Seq() => true;
      case Seq(nullVariable) => {
        if (nullVariable == result) {
          val values = AbstractGenerator.domainFrom(v0, v1, { _ + _ });

          result.dom = IntDomain(values: _*);

        } else if (nullVariable == v0) {

          v0.dom = IntDomain(generateValues(result, v1): _*);

        } else if (nullVariable == v1) {

          v1.dom = IntDomain(generateValues(result, v0): _*);

        } else {

          throw new IllegalStateException();

        }
        true;
      }
      case _ => false;
    }) {

      addConstraint(
        if (result.dom.size == 1) {
          new Eq(true, v0, result.dom.firstValue, v1);
        } else if (v0.dom.size == 1) {
          new Eq(false, v1, v0.dom.firstValue, result);
        } else if (v1.dom.size == 1) {
          new Eq(false, v0, v1.dom.firstValue, result);
        } else {
          new Add(result, v0, v1);
        });
      true

    } else {
      false
    }
  }

  def generateValues(result: Variable, variable: Variable) =
    AbstractGenerator.domainFrom(result, variable, { _ - _ })

}
